packages/common/src/util.js

```diff
@@ -63,3 +82,137 @@ export function normalizeOauthConfig(configuration) {
 
   return configuration;
 }
+
+const separateUrl = fullUrl => {
+  const url = new URL(fullUrl);
+  return {
+    baseUrl: url.origin,
+    path: url.pathname,
+  };
+};
+
+/**
+ * The `defaultOptions` constant is an object that defines default values for various options that can
+ * be passed to the `request` function. These options include
+ * `headers`,
+ * `query`,
+ * `timeout`,
+ * `tls`,
+ * `body`, and
+ * `errors`.
+ */
+const defaultOptions = {
+  timeout: 300, // Default to 300 seconds
+  headers: {},
+  query: undefined,
+  body: undefined,
+  errors: {},
+  tls: {},
+};
+
+const assertOK = (response, errorMap, fullUrl) => {
+  const errMapMessage = errorMap[response.statusCode];
+
+  if (errMapMessage || response.statusCode >= 400) {
+    const defaultErrorMesssage = `Request to ${fullUrl} failed with status: ${response.statusCode}`;
+
+    const errMessage =
+      typeof errMapMessage === 'function'
+        ? errMapMessage(response)
+        : errMapMessage || defaultErrorMesssage;
+
+    const error = new Error(errMessage);
+    error.code = response.statusCode;
+    error.url = fullUrl;
+    throw error;
+  }
+};
+
+const isValidHttpUrl = fullUrl => {
+  try {
+    const { protocol } = new URL(fullUrl);
+    return protocol === 'http:' || protocol === 'https:';
+  } catch (_) {
+    return false;
+  }
+};
+
+/**
+ * The function `request` is an asynchronous function that sends HTTP requests and returns the response
+ * data, headers, and status code.
+ * @param method - The HTTP method to use for the request (e.g., "GET", "POST", "PUT", "DELETE", etc.).
+ * @param fullUrl - The full URL is the complete URL of the request, including the protocol (e.g.,
+ * "http://example.com/api").
+ * @param [options] - The `options` parameter is an object that contains additional configuration
+ * options for the request. It can have the following properties:
+ * @returns an object with the following properties:
+ * - code: the status code of the response
+ * - headers: the headers of the response
+ * - data: the body of the response
+ */
+export async function request(method, fullUrlOrPath, options = {}) {
+  let baseUrl, path;
+
+  if (!options.baseUrl || isValidHttpUrl(fullUrlOrPath)) {
+    ({ baseUrl, path } = separateUrl(fullUrlOrPath));
+  } else {
+    baseUrl = options.baseUrl;
+    path = fullUrlOrPath;
+  }
+
+  const { headers, query, body, errors, timeout, tls } = {
+    ...defaultOptions,
+    ...options,
+  };
+
+  if (!headers['Content-Type'] && typeof body !== 'string') {
+    headers['Content-Type'] = 'application/json';
+  }
+
+  const client = getClient(baseUrl);
+
+  const response = await client.request({
+    path,
+    query,
+    method,
+    headers,
+    body: body ? JSON.stringify(body) : undefined,
```

Comment:
hmm, I think body parsing will be a bit more complicated than this because we may not always be posting json. Maybe its ok to leave it for now and come back to it later.